### PR TITLE
segger-jlink: init at 7.20

### PR DESCRIPTION
**Pull request by FlorianFranzen**
###### Motivation for this change

The SEGGER JLink Software package was missing from nixpkgs.

I was able to test the 64- and 32-bit version on my local machine. I tried to test both ARM version with binfmt, with mixed results. I was able to build the `aarch64` version, but it seems `autoPatchelf` does not detect any of the dynamic libraries (even though just running `patchelf --print-needed <bin>` displays them correctly. The `armv7l`requires to build 400+ dependencies not in the binary cache, which is slow under binfmt, so I did not have a chance to check that yet.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by evils** [comment]
#80990 suggests this could be made available for x86 and ARM, 32 and 64 bit

**Comment by FlorianFranzen** [comment]
> #80990 suggests this could be made available for x86 and ARM, 32 and 64 bit

Indeed, and the way it makes udev available is much nicer as well. I will see how we can combine our approaches.

**Comment by FlorianFranzen** [comment]
@reardencode Thank you for your work in #80990, I took over a lot of your changes into my derivation and cited you as co-author. Would love to hear what you think about this.

@prusnak @liff @mtetreault I saw you looked in this in #80990, any feedback and testing would be greatly appreciated.

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -8285,6 +8285,12 @@
     githubId = 1973389;
     name = "Reuben D'Netto";
   };
+  reardencode = {
```
Comment (evils) [inline maintainers/maintainer-list.nix:8288]: changes to maintainer-list.nix should be separate commits

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 2]
```diff
@@ -0,0 +1,118 @@
+{ lib
+, stdenv
+, fetchurl
+, autoPatchelfHook
+, makeWrapper
+, qt4
+, libudev
```
Comment (evils) [inline pkgs/development/tools/misc/segger-jlink/default.nix:7]: libudev is an alias of udev, the library is in udev.dev, use udev and the .dev output should be used by nix for building

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    configuration.nix:
+      nixpkgs.config.allowUnfree = true;
+      nixpkgs.config.segger-jlink.acceptLicense = true;
+
+    config.nix:
+      allowUnfree = true;
+      segger-jlink.acceptLicense = true;
+
+    [1]: https://www.segger.com/downloads/jlink/JLink_Linux_V720_x86_64.tgz
```
Comment (evils) [inline pkgs/development/tools/misc/segger-jlink/default.nix:54]: minor nitpick, but filling in ${system} here could facilitate them having different terms for different systems
and avoid minor confusion for non-x86_64 users

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    [2]: https://www.segger.com/purchase/licensing/
+  '';
+
+in stdenv.mkDerivation {
+  pname = "segger-jlink";
+  version = "7.20";
```
Comment (mtetreault) [inline pkgs/development/tools/misc/segger-jlink/default.nix:60]: You could use  720 as the version name, you could then use a variable in the url.

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  nativeBuildInputs = [ autoPatchelfHook ];
+  buildInputs = [ qt4 libudev ];
+
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    # Install all files to JLink first
+    mkdir -p $out/JLink
```
Comment (mtetreault) [inline pkgs/development/tools/misc/segger-jlink/default.nix:80]: You could create both folder at the same time.

```
$ mkdir -p $out/{Jlink,bin}

**Review by mtetreault — COMMENTED** [review]
Thanks for submitting this, I had this on my todo list for some times already. Nice works, see comment below.

**Comment by liff** [comment]
Hi, I was writing a flake for this at some point, with an intent to later submit it to nixpkgs. I pushed what I have to a [repo](https://github.com/liff/j-link-flake/blob/main/pkgs/j-link/default.nix).

You could consider adding the [desktop items](https://github.com/liff/j-link-flake/blob/main/pkgs/j-link/default.nix#L39-L114), which I “stole” from the Windows/macOS packages. Unfortunately the Linux packages do not come with icon(s) for the items. There’s one in Windows and macOS package but it’s generic and low resolution.

Another issue is that you may want to prefer the bundled Qt4 libraries to the ones from nixpkgs. IIRC the UI looks a bit crappier, especially on HiDPI, when using qt4 from nixpkgs. Maybe Segger has modified their Qt or something 🤷🏻. At least check out if it makes any difference on your system, and on 7.20.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (FlorianFranzen) [reply maintainers/maintainer-list.nix:8288]: Done, good catch.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 2 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix]: Done and good to know. The `dev` output is automatically included for `buildInputs`, so it was a simple change.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 3 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix]: Yes, indeed. I added the url to the let, so it is only constructed once and used in the message and the fetchurl statement.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 4 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix]: I was a bit hesitant, because the website officially uses that form. But given that it is really the download link that does not specify this better, it would be a cosmetic thing at best, so I went with your approach in the end.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 5 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix:84]: I would like to keep each block easily relocatable for the future, as there is not guarantee to how the structure of the download will develop in the future. I am sure I could have combined various calls in this way, also by using `install` instead, but my main focus was really on keeping the install script easily readable.

Comment (mtetreault) [reply pkgs/development/tools/misc/segger-jlink/default.nix:84]: Alright, fair enough.

**Comment by reardencode** [comment]
@FlorianFranzen I'm sorry that I've been delinquent on my PR for JLink. Thanks for picking it up!

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      rm -r $out/JLink/{arm,lib*_arm.so*}
+    ''}
+
+    # Move docs and examples
+    mkdir -p $out/share
+    mv $out/JLink/Doc $out/share/docs
+    mv $out/JLink/Samples $out/share/examples
+
+    # Move udev rule
+    mkdir -p $out/lib/udev/rules.d
```
Comment (maxhbr) [inline pkgs/development/tools/misc/segger-jlink/default.nix:116]: Also the `.so` files should be exposed, so that for example `nrfjprog` can find them:
```suggestion
    mkdir -p $out/lib/udev/rules.d
    ln -s $out/JLink/*.so* $out/lib
```

**Comment by maxhbr** [comment]
This code works for me and with the above mentioned minor patch I am able to use it with nrfjprog (see https://github.com/maxhbr/myconfig/blob/f5910d557a22073b59897733e140d76b1a63ee66/flakes/zephyrproject/nRF-Command-Line-Tools/default.nix)  to flash the nrf5340-dk with Zephyr.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 6 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix:120]: Of course, no problem. I added the necessary commands. Thank you for the feedback.

**Comment by FlorianFranzen** [comment]
Added commands to symlink libraries as well as updated version and hashes to latest release.

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+    [1]: ${url}
+    [2]: https://www.segger.com/purchase/licensing/
+  '';
+
+in stdenv.mkDerivation {
+  pname = "segger-jlink";
+  inherit version;
+
+  src = assert !acceptLicense -> throwLicense; fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/segger-jlink/default.nix:66]:
```suggestion
  src = fetchurl {
```

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    };
+  };
+
+  platform = supported.${system} or (throw "unsupported platform ${system}");
+
+  version = "750";
+
+  url = "https://www.segger.com/downloads/jlink/JLink_Linux_V${version}_${platform.name}.tgz";
+
+  throwLicense = throw ''
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/segger-jlink/default.nix:39]:
```suggestion
  assert !acceptLicense -> throw ''
```

File: pkgs/development/tools/misc/segger-jlink/default.nix [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  dontBuild = true;
+
+  installPhase = ''
+    # Install all files to JLink first
+    mkdir -p $out/JLink
+    cp -r * $out/JLink
+
+    # Symlink all binaries into bin
+    mkdir -p $out/bin
+    for prog in $out/JLink/J*; do
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/segger-jlink/default.nix:89]: Does this leave files at ``$out/JLink`` behind? We cannot pollute / like this.

Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix:89]: Yes, indeed it does. There is some requirement to have some of these files relative to the executable. Where would you move them instead? `share/JLink` maybe?

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 8 above)
Comment (FlorianFranzen) [reply pkgs/development/tools/misc/segger-jlink/default.nix]: Thank you, added it in an extra commit for now. Will squash once we found a good way to move `JLink` folder.

File: pkgs/development/tools/misc/segger-jlink/default.nix (hunk 9 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/misc/segger-jlink/default.nix:89]: > `share/JLink` maybe?

sounds better than the root direcotry.

**Comment by liff** [comment]
I’m getting a syntax error when building:

```
error: syntax error, unexpected ASSERT

       at /tmp/nixpkgs/pkgs/development/tools/misc/segger-jlink/default.nix:39:3:

           38|
           39|   assert !acceptLicense -> throw ''
             |   ^
           40|     Use of the "SEGGER JLink Software and Documentation pack" requires the
```

Anyway, isn’t [`requireFile`](https://github.com/NixOS/nixpkgs/blob/3e0ce8c5d478d06b37a4faa7a4cc8642c6bb97de/pkgs/build-support/trivial-builders.nix#L471-L491) the usual way to handle this sort of license thing?

**Comment by StarGate01** [comment]
Superseded by and incorporated into by #178603 .
